src/battle/battle_setup.c

```diff
@@ -0,0 +1,575 @@
+    BattleSetup* setup = BattleSetup_New(heapId, BATTLE_TYPE_BUG_CONTEST);
+    setup->safariBalls = balls;
+    CopyPokemonToPokemon(bugmon, setup->bugContestMon);
+    return setup;
+}
+
+BattleSetup* BattleSetup_New_PalPark(HeapID heapId, int balls) {
+    BattleSetup* setup = BattleSetup_New(heapId, BATTLE_TYPE_PAL_PARK);
+    setup->safariBalls = balls;
+    return setup;
+}
+
+BattleSetup* BattleSetup_New_Tutorial(HeapID heapId, FieldSystem* fieldSystem) {
+    PlayerProfile* profile = Save_PlayerData_GetProfileAddr(fieldSystem->saveData);
+    OPTIONS* options = Save_PlayerData_GetOptionsAddr(fieldSystem->saveData);
+    BattleSetup* setup = BattleSetup_New(heapId, BATTLE_TYPE_TUTORIAL);
+    setup->saveData = fieldSystem->saveData;
+    {
+    MsgData* msgData = NewMsgDataFromNarc(MSGDATA_LOAD_LAZY, NARC_msgdata_msg, NARC_msg_msg_0445_bin, heapId);
+    {
+    String* name = String_New(PLAYER_NAME_LENGTH + 1, heapId);
+    ReadMsgDataIntoString(msgData, PlayerProfile_GetTrainerGender(profile) ^ 1, name);
+    Save_Profile_PlayerName_Set(setup->profile[0], String_cstr(name));
+    String_Delete(name);
+    }
+    DestroyMsgData(msgData);
+    }
+    PlayerProfile_SetTrainerGender(setup->profile[0], PlayerProfile_GetTrainerGender(profile) ^ 1);
+    sub_02052504(setup, fieldSystem);
+    Options_Copy(options, setup->options);
+    setup->timeOfDay = Field_GetTimeOfDay(fieldSystem);
+    Bag_AddItem(setup->bag, ITEM_POKE_BALL, 20, heapId);
+    {
+    Pokemon* pokemon = AllocMonZeroed(heapId);
+    CreateMon(pokemon, SPECIES_MARILL, 5, 32, FALSE, 0, OT_ID_RANDOM_NO_SHINY, 0);
+    Party_AddMon(setup->party[0], pokemon);
+    CreateMon(pokemon, SPECIES_RATTATA, 2, 32, FALSE, 0, OT_ID_RANDOM_NO_SHINY, 0);
+    Party_AddMon(setup->party[1], pokemon);
+    FreeToHeap(pokemon);
+    }
+    setup->unk1CC[0] = 0;
+    setup->storagePC = SaveArray_PCStorage_Get(fieldSystem->saveData);
+    setup->unk_10C = fieldSystem->unk94;
+    setup->unk1B8 = NULL;
+    setup->gameStats = Save_GameStats_Get(fieldSystem->saveData);
+    setup->mapNumber = fieldSystem->location->mapId;
+    sub_02052580(setup);
+    return setup;
+}
+
+void BattleSetup_Delete(BattleSetup* setup) {
+    int i;
+
+    for (i = 0; i < 4; ++i) {
+        GF_ASSERT(setup->party[i] != NULL);
+        FreeToHeap(setup->party[i]);
+    }
+    
+    for (i = 0; i < 4; ++i) {
+        GF_ASSERT(setup->profile[i] != NULL);
+        FreeToHeap(setup->profile[i]);
+    }
+    
+    for (i = 0; i < 4; ++i) {
+        GF_ASSERT(setup->chatot[i] != NULL);
+        FreeToHeap(setup->chatot[i]);
+    }
+    
+    FreeToHeap(setup->bag);
+    FreeToHeap(setup->pokedex);
+    FreeToHeap(setup->options);
+    sub_02067A78(setup->unk_134);
+    FreeToHeap(setup->bugContestMon);
+    FreeToHeap(setup);
+}
+
+void BattleSetup_AddMonToParty(BattleSetup* setup, Pokemon* mon, int battlerId) {
+    GF_ASSERT(battlerId < 4);
```

Comment:
Add a `BATTLER_MAX` variable to `constants/battle.h`, then `battlerId < BATTLER_MAX`